js/search.js

```diff
@@ -0,0 +1,122 @@
+
+// Retrieves the value of a GET parameter with a given key
+// Accepts:
+//  param: string
+// Returns:
+//  string or null
+var getParam = function(param) {
+  var queryString = window.location.search.substring(1), queries = queryString.split('&');
+  for (var i in queries) {
+    var pair = queries[i].split('=');
+    if (pair[0] === param) {
+      // Decode the parameter value, replacing %20 with a space etc.
+      return decodeURI(pair[1]);
+    }
+  }
+  return null;
+};
+
+  // Filters posts with the condition `post['property'] == value`
+  // Accepts:
+  //  posts - array of post objects and a string
+  //  property - string of post object property to compare
+  //  value - filter value of property
+  // Returns:
+  //  array of post objects
+var filterPostsByPropertyValue = function(posts, property, value) {
+  var filteredPosts = [];
+  // The last element is a null terminator
+  posts.pop();
+  for (var i in posts) {
+    var post = posts[i],
+    prop = post[property];
+
+    // Last element of tags is null
+    post.tags.pop();
+
+    // The property could be a string, such as a post's category,
+    // or an array, such as a post's tags
+    if (prop.constructor === String) {
+      if (prop.toLowerCase() === value.toLowerCase()) {
+        filteredPosts.push(post);
+      }
+    } else if (prop.constructor === Array) {
+      for (var j in prop) {
+        if (prop[j].toLowerCase() === value.toLowerCase()) {
+          filteredPosts.push(post);
+        }
+      }
+    }
+  }
+  return filteredPosts;
+};
+
+// Formats search results and appends them to the DOM
+// Accepts:
+//  property: string of object type we're displaying
+//  value: string of name of object we're displaying
+//  posts: array of post objects
+// Returns:
+//  undefined
+var layoutResultsPage = function(property, value, posts) {
+  var dict = {};
+  dict["posts"] = posts;
+  dict["baseurl"] = "";
+
+  console.log("Going to display: ", dict);
+  getTemplateAjax("search_results_list.html", "#results", dict, displayTemplate);
+};
+
+function getTemplateAjax(path, element, dict, callback) {
+  $.ajax({
+    url: path,
+    success: function(data) {
+      //Get the Template and compile it
+      var template  = Handlebars.compile(data);
+      if (callback) {
+        callback(template, element, dict);
+      }
+    }
+  });
+}
+
+var displayTemplate = function(template, element, data) {
+  //Replace the body section with the new code.
+  $(element).append(template(data));
+}
+
+// Formats the search results page for no results
+// Accepts:
+//  property: string of object type we're displaying
+//  value: string of name of object we're displaying
+// Returns:
+//  undefined
+var noResultsPage = function(property, value) {
+  console.log ("No results found");
+  $('#results').html(
+    '<br/><br/>\
+    <h1>No Results Found</h1>\
+    <p>Sorry, none of these recipes are described as ‘' + value + '’.</p>'
+  );
```

Comment:
This introduces a reflected XSS vulnerability - the user can control `value` and set it to `<img src='/nothing' onerror='alert("Running JS")' />`. While this isn't severe at the moment, since the site doesn't have sessions, it would be better to fix it.

Putting this HTML into the page by default, in a hidden div, and showing that div when no results are found, is a safer way of doing this. You can set the value you want on the page by adding a span, then referencing that span by ID and setting it's .text (which is sanitized, where .html isn't).

This might look something like this:
`$('#results').show();
$('#results .value').text(value);`